src/cli/commands/pack.js

```diff
@@ -68,14 +68,17 @@ function addEntry(packer: any, entry: Object, buffer?: ?Buffer): Promise<void> {
 
 export async function pack(config: Config, dir: string): Promise<stream$Duplex> {
   const pkg = await config.readRootManifest();
-  const {bundledDependencies, files: onlyFiles} = pkg;
+  const {bundledDependencies, main, files: onlyFiles} = pkg;
 
   // inlude required files
   let filters: Array<IgnoreFilter> = NEVER_IGNORE.slice();
   // include default filters unless `files` is used
   if (!onlyFiles) {
     filters = filters.concat(DEFAULT_IGNORE);
   }
+  if (main) {
```

Comment:
Hmm, still not sure I understand. In `__tests__/fixtures/pack/files-include-mandatory/package.json`, the `main` field is already `index.js`. The point of this change is to make sure that `index.js` *doesn't* have to be in `files` in order for the packed tarball to contain `index.js`.

I split my changes into two commits - test first, fix second - which should help demonstrate how the new condition is covered by the test. If you run the following at each commit, you'll see how the tarball changes (this is also covered by `yarn test`, but I pushed the two commits too quickly, and the CI for the first commit seems to have run against the changes in the second commit):

```
yarn build && cd __tests__/fixtures/pack/files-include-mandatory && ../../../../bin/yarn.js pack && tar tvf *tgz | grep index.js; cd -
```